### PR TITLE
Semantic Deduplication Documentation

### PR DESCRIPTION
**Pull request by VibhuJawa**
## Description

This PR adds Semantic Deduplication Documentation.

### CODE REVIEW

File: docs/user-guide/semdedup.rst [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+Usage
+-----------------------------------------
+
+Before running semantic deduplication, ensure that each document in your dataset has a unique identifier. You can use the `add_id` module from NeMo Curator if needed:
+
+.. code-block:: bash
+
+    add_id \
+      --input-data-dir=<Path to directory containing jsonl files> \
+      --log-dir=./log/add_id
```
Comment (sarahyurick) [inline docs/user-guide/semdedup.rst:116]: Is there no API support for this yet?

File: docs/user-guide/semdedup.rst [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+Performance Considerations
+-----------------------------------------
+
+Semantic deduplication is computationally intensive, especially for large datasets. However, the benefits in terms of reduced training time and improved model performance often outweigh the upfront cost. Consider the following:
+
+- Use GPU acceleration for faster embedding generation and clustering.
+- Adjust the number of clusters (`n_clusters`) based on your dataset size and available computational resources.
+- The `eps_to_extract` parameter allows you to control the trade-off between dataset size reduction and potential information loss.
+
+For more details on the algorithm and its performance implications, refer to the original paper: "SemDeDup: Data-efficient learning at web-scale through semantic deduplication" by Abbas et al.
```
Comment (sarahyurick) [inline docs/user-guide/semdedup.rst:232]: Link again.

**Review by sarahyurick — CHANGES_REQUESTED** [review]
The steps are very clear, thanks! Just added a few minor comments.

File: docs/user-guide/semdedup.rst (hunk 2 above)
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst]: Resolved by: https://github.com/NVIDIA/NeMo-Curator/pull/159/commits/5eae8a68096518582c23fe6d1336224b31112b45

File: docs/user-guide/semdedup.rst [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+#######################################################
+
+-----------------------------------------
+Background
+-----------------------------------------
+
+Semantic deduplication is an advanced technique for removing redundant data from large datasets by identifying and eliminating semantically similar documents.
+Unlike exact or fuzzy deduplication, which focus on textual similarity, semantic deduplication leverages the semantic meaning of the content to identify duplicates.
+
+As outlined in the paper "SemDeDup: Data-efficient learning at web-scale through semantic deduplication" by Abbas et al.,
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst]: Resolved by : https://github.com/NVIDIA/NeMo-Curator/pull/159/commits/5eae8a68096518582c23fe6d1336224b31112b4

**Review by ryantwolf — COMMENTED** [review]
Looks good overall. Main thing is to emphasize how this can be used for multiple modalities and limit the number of times we reference text-specific stuff.

File: docs/user-guide/semdedup.rst [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+#######################################################
+Semantic Deduplication
+#######################################################
+
+-----------------------------------------
+Background
+-----------------------------------------
+
+Semantic deduplication is an advanced technique for removing redundant data from large datasets by identifying and eliminating semantically similar documents.
+Unlike exact or fuzzy deduplication, which focus on textual similarity, semantic deduplication leverages the semantic meaning of the content to identify duplicates.
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst:12]: Ha ha, I added it and removed it because we dont have an example yet. Let me add it back.

**Review by sarahyurick — APPROVED** [review]
LGTM!

File: docs/user-guide/semdedup.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst:12]: Resolved

File: docs/user-guide/semdedup.rst [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+ Semantic deduplication is particularly effective for large, uncurated web-scale datasets, where it can remove up to 50% of the data with minimal performance loss.
+The semantic deduplication module in NeMo Curator uses embeddings from to identify and remove "semantic duplicates" - data pairs that are semantically similar but not exactly identical.
+
+-----------------------------------------
+How It Works
+-----------------------------------------
+
+The SemDeDup algorithm consists of the following main steps:
+
+1. Embedding Generation: Each data point is embedded using a pre-trained language model.
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst]: Resolved by https://github.com/NVIDIA/NeMo-Curator/pull/159/commits/dfdc62869479e80cb91b78bca5ad15352cfda53b

File: docs/user-guide/semdedup.rst [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+To perform semantic deduplication, you can either use individual components or the SemDedup class with a configuration file:
+
+Using individual components:
+
+1. Embedding Creation:
+
+.. code-block:: python
+
+    from nemo_curator.modules.semantic_dedup import EmbeddingCreator
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst]: Resolved by https://github.com/NVIDIA/NeMo-Curator/pull/159/commits/dfdc62869479e80cb91b78bca5ad15352cfda53b

File: docs/user-guide/semdedup.rst [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+- `eps_to_extract`: Deduplication threshold. Higher values result in more aggressive deduplication.
+- `which_to_keep`: Strategy for choosing which duplicate to keep ("hard" or "soft").
+
+-----------------------------------------
+Output
+-----------------------------------------
+
+The semantic deduplication process produces a deduplicated dataset, typically reducing the dataset size by 20-50% while maintaining or improving model performance. The output includes:
+
+1. Embeddings for each document
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst]: Resolved by https://github.com/NVIDIA/NeMo-Curator/pull/159/commits/dfdc62869479e80cb91b78bca5ad15352cfda53b

File: docs/user-guide/semdedup.rst [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+How It Works
+-----------------------------------------
+
+The SemDeDup algorithm consists of the following main steps:
+
+1. Embedding Generation: Each data point is embedded using a pre-trained language model.
+2. Clustering: The embeddings are clustered into k clusters using k-means clustering.
+3. Similarity Computation: Within each cluster, pairwise cosine similarities are computed.
+4. Duplicate Identification: Data pairs with cosine similarity above a threshold are considered semantic duplicates.
+5. Duplicate Removal: From each group of semantic duplicates within a cluster, one representative document is kept (typically the one with the lowest cosine similarity to the cluster centroid) and the rest are removed.
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst]: Resolved by https://github.com/NVIDIA/NeMo-Curator/pull/159/commits/dfdc62869479e80cb91b78bca5ad15352cfda53b

File: docs/user-guide/semdedup.rst [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    max_iter: 100
+    kmeans_with_cos_dist: false
+
+    # Semdedup configuration
+    which_to_keep: "hard"
+    largest_cluster_size_to_process: 100000
+    sim_metric: "cosine"
+
+    # Extract dedup configuration
+    eps_thresholds:
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst:66]: Resolved by: https://github.com/NVIDIA/NeMo-Curator/pull/159/commits/33634a94c471236fce74f1c539a9df1a091e45ff

File: docs/user-guide/semdedup.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply docs/user-guide/semdedup.rst:12]: Done

**Comment by VibhuJawa** [comment]
@ryantwolf , Please feel free to review the PR again. Have made suggested changes

**Review by ryantwolf — APPROVED** [review]
Excellent!

**Pull request merged**